packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -0,0 +1,181 @@
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import "@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+
+import "@opengsn/contracts/src/forwarder/IForwarder.sol";
+import "@opengsn/contracts/src/BaseRelayRecipient.sol";
+import "@opengsn/contracts/src/BasePaymaster.sol";
+import "@opengsn/contracts/src/utils/GsnUtils.sol";
+
+import "./interfaces/IChainlinkOracle.sol";
+
+import "./helpers/UniswapV3Helper.sol";
+import "./helpers/PermitInterfaceDAI.sol";
+
+/**
+ * A paymaster allowing addresses holding ERC20 tokens with 'permit' functionality
+ * to pay for a GSN transaction.
+ */
+contract PermitERC20UniswapV3Paymaster is BasePaymaster, BaseRelayRecipient {
+    using SafeMath for uint256;
+
+    event Received(address indexed sender, uint256 eth);
+    event TokensCharged(uint256 gasUseWithoutPost, uint256 gasJustPost, uint256 ethActualCharge);
+
+    IChainlinkOracle public priceFeed;
+    ISwapRouter public uniswap;
+    IERC20 public token;
+    IERC20 public weth;
+
+    uint24 public uniswapPoolFee;
+    uint256 public gasUsedByPost;
+    bytes4 public permitMethodSignature;
+    uint256 public permitMethodDataLength;
+
+    function versionPaymaster() external override virtual view returns (string memory){
+        return "2.2.3+opengsn.permit-erc20-uniswap-v3.ipaymaster";
+    }
+
+    function versionRecipient() external override virtual view returns (string memory){
+        return "2.2.3+opengsn.permit-erc20-uniswap-v3.irelayrecipient";
+    }
+
+    function trustedForwarder() override(BasePaymaster, BaseRelayRecipient) public view returns (address forwarder){
+        forwarder = BaseRelayRecipient.trustedForwarder();
+    }
+
+    function setTrustedForwarder(address _forwarder) override(BasePaymaster, BaseRelayRecipient) public onlyOwner {
+        BaseRelayRecipient.setTrustedForwarder(_forwarder);
+    }
+
+    function _msgSender() internal view override(Context, BaseRelayRecipient) returns (address payable sender) {
+        sender = BaseRelayRecipient._msgSender();
+    }
+
+    function _msgData() internal view override(Context, BaseRelayRecipient) returns (bytes memory) {
+        return BaseRelayRecipient._msgData();
+    }
+
+    constructor(
+        IERC20 _weth,
+        IERC20 _token,
+//        IQuoter _quoter,
+        IRelayHub _relayHub,
+        ISwapRouter _uniswap,
+        IChainlinkOracle _priceFeed,
+        address _trustedForwarder,
+        uint24 _uniswapPoolFee,
+        uint256 _gasUsedByPost,
+        uint256 _permitMethodDataLength,
+        string memory _permitMethodSignature
+    ) {
+        weth = _weth;
+        token = _token;
+//        quoter = _quoter;
+        uniswap = _uniswap;
+        priceFeed = _priceFeed;
+        uniswapPoolFee = _uniswapPoolFee;
+        permitMethodSignature = bytes4(keccak256(bytes(_permitMethodSignature)));
```

Comment:
ok, that's neat. I thought of enabling multiple signatures, but this is very explicit.

Reply:
constructor-assigned members should be immutable.